src/test/java/com/microsoft/azure/relay/SendReceiveTest.java

```diff
@@ -172,147 +169,148 @@ public void websocketMultipleSenderTest() throws URISyntaxException {
+		listener.setRequestHandler(context -> sendResponseMessages(context, LARGE_BYTES, SMALL_BYTES));
+		sendHttpRequest(TestUtil.concatByteArrays(LARGE_BYTES, SMALL_BYTES), SMALL_BYTES);
 	}
 	
 	@Test
 	public void httpWriteResponseAfterFlushTimerTest() throws IOException {
-		listener.setRequestHandler(context -> sendResponseMessages(context, new String[] {smallStr, smallStr}, true));
-		httpRequestSender("POST", smallStr + smallStr, smallStr);
-	}
-	
-	@Test
-	public void httpWriteResponseEnsureOrderTest() throws IOException {
-		String[] messages = new String[1010]; // 1000 small messages followed by 10 large ones
-		for (int i = 0; i < messages.length; i++) {
-			messages[i] = i + ((i < 1000) ? smallStr : largeStr);
-		}
-		listener.setRequestHandler(context -> sendResponseMessages(context, messages, false));
-		httpRequestSender("POST", String.join("", messages), smallStr);
+		listener.setRequestHandler(context -> sendResponseMessages(context, true, SMALL_BYTES, SMALL_BYTES));
+		sendHttpRequest(TestUtil.concatByteArrays(SMALL_BYTES, SMALL_BYTES), SMALL_BYTES);
 	}
 	
-	private static CompletableFuture<Void> websocketClient(String msgExpected, String msgToSend) {
+	private static CompletableFuture<Void> sendAndReceiveWithWebsocketClient(byte[] msgExpected, byte[] msgToSend) {
 		AtomicBoolean receivedReply = new AtomicBoolean(false);
 		
-		return client.createConnectionAsync().thenCompose((channel) -> {
-			return channel.writeAsync(StringUtil.toBuffer(msgToSend)).thenCompose($void -> {
+		return client.createConnectionAsync().thenCompose(channel -> {
+			return channel.writeAsync(ByteBuffer.wrap(msgToSend.clone())).thenCompose($void -> {
 				return channel.readAsync();
-			}).thenCompose((bytesReceived) -> {
-				String msgReceived = new String(bytesReceived.array());
+			}).thenCompose(bytesReceived -> {
+				byte[] msgReceived = bytesReceived.array();
 				receivedReply.set(true);
-				assertEquals("Websocket sender did not receive the expected reply.", msgExpected, msgReceived);
+				assertTrue("Websocket sender did not receive the expected reply.", Arrays.equals(msgExpected, msgReceived));
 				return channel.closeAsync();
 			}).thenRun(() -> assertTrue("Did not receive message from websocket sender.", receivedReply.get()));
 		});
 	}
 	
-	private static CompletableFuture<Void> websocketListener(String msgExpected, String msgToSend) {
-		return listener.acceptConnectionAsync().thenComposeAsync((channel) -> {
+	private static CompletableFuture<Void> sendAndReceiveWithWebsocketListener(byte[] msgExpected, byte[] msgToSend) {
+		return listener.acceptConnectionAsync().thenComposeAsync(channel -> {
 			return channel.readAsync().thenAccept(bytesReceived -> {
-				assertEquals("Websocket listener did not receive the expected message.", msgExpected, new String(bytesReceived.array()));
+				assertTrue("Websocket listener did not receive the expected message.", Arrays.equals(msgExpected, bytesReceived.array()));
```

Comment:
>bytesReceived.array() [](start = 102, length = 21)

If the ByteBuffer ever changes to not happen to give you a perfectly sized array this will fail, right?

Reply:
It would fail, but currently `readAsync` returns a perfectly wrapped ByteBuffer. It's not like we are letting the user bring their own buffer anyways as of now

---
In reply to: [295927555](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295927555) [](ancestors = 295927555)